proxy/IPAllow.cc

```diff
@@ -81,47 +88,70 @@ IpAllow::reconfigure()
 IpAllow *
 IpAllow::acquire()
 {
-  return (IpAllow *)configProcessor.get(configid);
+  return static_cast<IpAllow *>(configProcessor.get(configid));
+}
+
+void
+IpAllow::release(IpAllow *config)
+{
+  configProcessor.release(configid, config);
 }
 
 void
-IpAllow::release(IpAllow *lookup)
+IpAllow::release()
+{
+  configProcessor.release(configid, this);
+}
+
+auto
+IpAllow::match(sockaddr const *ip, match_key_t key) -> ACL
 {
-  configProcessor.release(configid, lookup);
+  self_type *self = acquire();
+  void *raw       = nullptr;
+  if (SRC_ADDR == key) {
+    self->_src_map.contains(ip, &raw);
+    Record *r = static_cast<Record *>(raw);
+    // Special check - if checking in accept is enabled and the record is a deny all,
+    // then return a missing record instead to force an immediate deny. Otherwise it's delayed
+    // until after remap, to allow remap rules to tweak the result.
+    if (raw && r->_method_mask == 0 && r->_nonstandard_methods.empty() && accept_check_p) {
+      raw = nullptr;
+    }
+  } else {
+    self->_dst_map.contains(ip, &raw);
+  }
+  if (raw == nullptr) {
+    self->release();
+    self = nullptr;
+  }
+  return {static_cast<Record *>(raw), self};
 }
```

Comment:
So this is a constructor?

Reply:
Yes. The compiler knows the return type and so automatically coerces the initializer list in to the returned object type.